src/app/mesh/qgsmaptooleditmeshframe.cpp

```diff
@@ -2717,7 +2786,20 @@ void QgsMapToolEditMeshFrame::addVertex(
     zValue = QgsMeshLayerUtils::interpolateFromVerticesData( v1, v2, v3, v1.z(), v2.z(), v3.z(), mapPoint );
   }
   else
-    zValue = currentZValue();
+  {
+    if ( mZValueWidget->getZFromProjectElevation() )
+    {
+      const QgsAbstractTerrainProvider *terrainProvider = QgsProject::instance()->elevationProperties()->terrainProvider();
+      const QgsCoordinateTransform transformation = QgsCoordinateTransform( mCurrentLayer->crs(), terrainProvider->crs(), QgsProject::instance() );
+      const QgsPointXY point = transformation.transform( effectivePoint.x(), effectivePoint.y() );
+
+      zValue = terrainProvider->heightAt( point.x(), point.y() );
+    }
+    else
+    {
+      zValue = currentZValue();
+    }
+  }
```

Comment:
This branch should be tested first, before all the `else if`s.
The idea is that if _from terrain_ is enabled, this should take precedence.
Then if not, use current z value, or snap, or edge, or face.

Also the same _failed transformation_/NaN heightAt check should be performed here too, as above. ( maybe wrap that in a private method?)